cpp/src/aztec/honk/composer/composer_helper/permutation_helper.hpp

```diff
@@ -1,118 +1,89 @@
- * Compute wire copy cycles
- *
- * First set all wire_copy_cycles corresponding to public_inputs to point to themselves.
- * Then go through all witnesses in w_l, w_r, w_o and w_4 (if program width is > 3) and
- * add them to cycles of their real indexes.
+ * Compute all CyclicPermutations of the circuit. Each CyclicPermutation represents the indices of the values in the
+ * witness wires that must have the same value.
  *
  * @tparam program_width Program width
  * */
 template <size_t program_width, typename CircuitConstructor>
-void compute_wire_copy_cycles(CircuitConstructor& circuit_constructor, CycleCollector& wire_copy_cycles)
+std::vector<CyclicPermutation> compute_wire_copy_cycles(const CircuitConstructor& circuit_constructor)
 {
     // Reference circuit constructor members
-    const std::vector<uint32_t>& real_variable_index = circuit_constructor.real_variable_index;
-    const std::vector<uint32_t>& public_inputs = circuit_constructor.public_inputs;
-
-    const std::vector<uint32_t>& w_l = circuit_constructor.w_l;
-    const std::vector<uint32_t>& w_r = circuit_constructor.w_r;
-    const std::vector<uint32_t>& w_o = circuit_constructor.w_o;
-    const size_t n = circuit_constructor.n;
-    const std::vector<uint32_t>& w_4 = circuit_constructor.w_4;
-
-    size_t number_of_cycles = 0;
-
+    const size_t num_gates = circuit_constructor.num_gates;
+    std::span<const uint32_t> public_inputs = circuit_constructor.public_inputs;
     const size_t num_public_inputs = public_inputs.size();
 
-    // Initialize wire_copy_cycles of public input variables to point to themselves ( we could actually ignore this step
-    // for HONK because of the way we construct the permutation)
-    for (size_t counter = 0; counter < num_public_inputs; ++counter) {
-        size_t i = num_public_inputs - 1 - counter;
-        cycle_node left{ static_cast<uint32_t>(i), WireType::LEFT };
-        cycle_node right{ static_cast<uint32_t>(i), WireType::RIGHT };
+    // Get references to the wires containing the index of the value inside constructor.variables
+    // These wires only contain the "real" gate constraints, and are not padded.
+    std::array<std::span<const uint32_t>, program_width> wire_indices;
+    wire_indices[0] = circuit_constructor.w_l;
+    wire_indices[1] = circuit_constructor.w_r;
+    wire_indices[2] = circuit_constructor.w_o;
+    if constexpr (program_width > 3) {
+        wire_indices[3] = circuit_constructor.w_4;
+    }
 
-        const auto public_input_index = real_variable_index[public_inputs[i]];
-        if (static_cast<size_t>(public_input_index) >= number_of_cycles) {
-            wire_copy_cycles.resize(public_input_index + 1);
-        }
-        std::vector<cycle_node>& cycle = wire_copy_cycles[static_cast<size_t>(public_input_index)];
+    // Each variable represents one cycle
+    const size_t number_of_cycles = circuit_constructor.variables.size();
+    std::vector<CyclicPermutation> copy_cycles(number_of_cycles);
```

Comment:
Initializes the `copy_cycles` to be the same size as `variables` means that there is a 1-to-1 correspondence between the two.